src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs

```diff
@@ -167,20 +184,26 @@ public bool AreEqual(object x, object y, ref Tolerance tolerance, bool topLevelC
             if (object.ReferenceEquals(x, y))
                 return true;
 
+            if (!state.TopLevelComparison && state.HasCompared(x, y))
+                return false;
+            state.RecordComparison(x, y);
```

Comment:
What happens if you only check `state.HasCompared(x, y)`, effectively replacing `TopLevelComparison`?

Reply:
The call pattern strongly suggests to me something like this, in the same vein as `bool HashSet<>.Add`:

```cs
if (!state.RecordComparison(x, y)) return false;
```

Reply:
Now I'm worried about something. And it's connected to `List<Comparison>` not being an immutable stack. (I really dislike mutable state. It's hard to think about properly.)

What about this scenario?

```cs
var equalInstance1 = new[] { 1 };
var equalInstance2 = new[] { 1 };

var x = new[] { equalInstance1, equalInstance1 };
var y = new[] { equalInstance2, equalInstance2 };

Assert.True(equalityComparer.AreEqual(x, y, ref tolerance));
```

If I understand the code properly, `equalInstance1` and `equalInstance2` will be compared once and then the pair will be recorded. Then they will be compared a second time and the line above will return `false` when it should return `true`.

A fix for this would be to use ImmutableStack or ImmutableQueue instead of List, and each time you chain to a comparer that is a level deeper (vertically, not horizontally), you do `comparisons.Push(new Comparison(x, y))`. A singly-linked list would work well since we can't reference System.Collections.Immutable. This is my favorite approach because it means that `state` is completely immutable, just like the original `topLevelComparison` was, and super easy to think about.

Reply:
Oh good catch! When I was writing this I was mostly concerned with duplicates down the call stack, but I'd completely forgotten about allowing horizontal duplicates. I'm going to have to play around a bit more with this.